opcua/client/ua_client.py

```diff
@@ -110,7 +110,7 @@ def _receive(self):
         elif isinstance(msg, ua.Acknowledge):
             self._call_callback(0, msg)
         elif isinstance(msg, ua.ErrorMessage):
-            self.logger.warning("Received an error: %s", msg)
+            self._call_callback(0, ua.UaStatusCodeError(msg.Error.value))
```

Comment:
this might also be correct. but I am a bit worried of the consequences since it might change a lot of things

Reply:
Yea, I'm not too sure how to handle the `request_id` parameter for the [`_call_callback()` method](https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/client/ua_client.py#L61-L66) because it seemed like the `request_id` was created by the client for tracking message parsing. Reading the "Error handling" section of the OPC UA docs (Part 6, 7.1.5), an *Error Message* should close the *TransportConnection* (which sends the *Hello/Acknowledge Messages*). Since the *Acknowledge* message uses `0`, I thought it'd be safe to use `0` because if the hello returns an error, the error would take the place of the acknowledgement; and if some later request returned an error, setting `0` with the error could prompt the client to close the *TransportConnection* gracefully. For example, if the client gets an error while sending messages, it can save the `_callbackmap` to a temporary file for debugging. I think right now the client reacts to when the Server closes the connection and disconnects the socket, clearing the `_callbackmap`.

Reply:
I think you should re-enable the logging waring and in fact call logger.fatal as receiving such a message is fatal, as I understand from doc. In fact maybe we should even close the connection ourself at that point, not sure...

Reply:
@urianchang can you re-enable the warnings? or did you have a very good reason to remove them? see comments above

Reply:
I've re-enabled the warnings (and elevated this one to fatal).

Reply:
@oroulet, sorry about the delayed response!

Reply:
@urianchang I was even more late... I forgot about that one...